tyrian/js/src/main/scala/tyrian/runtime/TyrianRuntime.scala

```diff
@@ -47,63 +47,92 @@ final class TyrianRuntime[F[_]: Concurrent, Model, Msg](
   @SuppressWarnings(Array("scalafix:DisableSyntax.var"))
   private var aboutToRunSubscriptions: Set[String] = Set.empty
 
-  def onMsg(run: RunWithCallback[F, Msg])(msg: Msg): Unit =
-    val (updatedState: Model, cmd: Cmd[F, Msg]) = update(msg, currentState)
-    currentState = updatedState
-    vnode = render(vnode, currentState, run)
-    performSideEffects(cmd, subscriptions(currentState), onMsg(run), run)
+  def initialise(cmd: Cmd[F, Msg]): Unit =
+    val res: F[Unit] =
+      for {
+        currentModel <- model.get
+        complete     <- completeUpdate(cmd, currentModel)
+      } yield complete
+
+    dispatcher.unsafeRunAndForget(res)
+
+  def onMsg(msg: Msg): Unit =
+    val res: F[Unit] =
+      for {
+        currentModel <- model.get
+        updated      <- Async[F].delay(update(msg, currentModel))
+        updatedState <- Async[F].delay(updated._1)
+        cmd          <- Async[F].delay(updated._2)
+        complete     <- completeUpdate(cmd, updatedState)
+      } yield complete
+
+    dispatcher.unsafeRunAndForget(res)
+
+  def completeUpdate(cmd: Cmd[F, Msg], updatedState: Model): F[Unit] =
+    val results: F[Stream[F, F[Unit]]] =
+      for {
+        _ <- model.set(updatedState)
+        n <- vnode.get
+        _ <- vnode.set(n match {
+          case Some(existingNode) => Some(render(existingNode, updatedState))
+          case None               => Some(render(node, updatedState))
+        })
+        sideEffects <- gatherSideEffects(cmd, subscriptions(updatedState))
+      } yield Stream.emits(sideEffects)
+
+    Async[F].flatMap(results) { (stream: Stream[F, F[Unit]]) =>
+      stream.foreach(channel.send(_).map(_ => ())).compile.drain
```

Comment:
`.map(_ => ())` can be replaced by `.void` (comes from the `Functor` syntax).

Reply:
Neat! I've changed that now.